### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Malox10**
Changed that users need to provide a valid email and password combination to login.

### CODE REVIEW

File: client/src/main/kotlin/frontend/MainView.kt [hunk 1]
```diff
@@ -3,26 +3,24 @@ package frontend
...
-	private val username: String
 	private val chatPath = SimpleListProperty(observableListOf(NodeCache.root))
 	private val comments = SimpleListProperty(observableListOf(NodeCache.root.children))
 
 	init {
-		val dialog = TextInputDialog().apply {
-			headerText = "please choose a username"
-			graphic = null
+		var authenticated = false
+		while (!authenticated) { val dialogResult = LoginDialog().showAndWait()
```
Comment (LastExceed) [inline client/src/main/kotlin/frontend/MainView.kt]: use do-while and fix formatting

File: server/src/main/kotlin/database/Database.kt [hunk 2]
```diff
@@ -28,10 +28,13 @@ class Database(private val db: ExposedDatabase) {
 			addLogger(StdOutSqlLogger)
 			SchemaUtils.create(TableSnapshot, TableNode, TableUser)
 			if (!TableNode.exists { TableNode.id eq 1L }) {//TODO: dont hardcode root id
-				val rootUserId = getOrCreateUser("rootuser")
-				createNode(rootUserId, "root of all evil", null)
+				val rootUserId = createUser("rootuser", "1234")
```
Comment (LastExceed) [inline server/src/main/kotlin/database/Database.kt]: "rootuser" should be an email address for consistency

File: server/src/main/kotlin/Server.kt [hunk 3]
```diff
@@ -39,15 +38,13 @@ class Server {
 			return
 		}
 
-		if (reader.readByte() != PacketId.NAMECHANGE.value) {
-			error("first packet must be a namechange")
-		}
-		val nameChange = reader.readPacketNameChange()
+		val login = handleLogin(reader, writer) ?: return client.dispose()
```
Comment (LastExceed) [inline server/src/main/kotlin/Server.kt:41]: inline handleLogin as this is the only occurrence

File: server/src/main/kotlin/Account.kt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+data class Account(
```
Comment (LastExceed) [inline server/src/main/kotlin/Account.kt]: unused class

File: server/src/main/kotlin/database/Database.kt (hunk 2 above)
Comment (LastExceed) [inline server/src/main/kotlin/database/Database.kt]: unused variable

File: client/src/main/kotlin/Networker.kt [hunk 5]
```diff
@@ -21,7 +21,10 @@ object Networker {
 		if (!response) {
 			error("outdated")
 		}
-		writer.writePacket((PacketNameChange(username)))
+		return login(email, password)
```
Comment (LastExceed) [inline client/src/main/kotlin/Networker.kt:24]: inline function call as its the only occurrence

**Review by LastExceed — CHANGES_REQUESTED** [review]

File: server/src/main/kotlin/database/Database.kt [hunk 6]
```diff
@@ -58,20 +61,21 @@ class Database(private val db: ExposedDatabase) {
...
-					it[name] = userName
-				} get TableUser.id
-			query[TableUser.id]
+	fun getUser(email: String, password: String): UserId? {
+		val userId = transaction(db) {
+			val row = TableUser.select { (TableUser.email eq email) and (TableUser.password eq password) }.firstOrNull()
+				?: return@transaction null
+			row[TableUser.id]
 		}
+		return if(userId == null) null else UserId(userId)
```
Comment (LastExceed) [inline server/src/main/kotlin/database/Database.kt]: simplify to `return userId?.let { UserId(it) }`

File: server/src/main/kotlin/Server.kt (hunk 3 above)
Comment (Malox10) [reply server/src/main/kotlin/Server.kt:41]: inlining would reduce testatibilty, abstraction, and understandability of the code.

File: server/src/main/kotlin/database/Database.kt [hunk 7]
```diff
@@ -28,9 +28,8 @@ class Database(private val db: ExposedDatabase) {
 			addLogger(StdOutSqlLogger)
 			SchemaUtils.create(TableSnapshot, TableNode, TableUser)
 			if (!TableNode.exists { TableNode.id eq 1L }) {//TODO: dont hardcode root id
-				val rootUserId = createUser("rootuser", "1234")
```
Comment (LastExceed) [inline server/src/main/kotlin/database/Database.kt]: this function call is necessary to populate a newly created database

**Review by LastExceed — CHANGES_REQUESTED** [review]

**Pull request merged**